src/platform/Darwin/DnssdImpl.cpp

```diff
@@ -69,259 +85,125 @@ std::string GetFullTypeWithSubTypes(const char * type, DnssdServiceProtocol prot
+MdnsContexts MdnsContexts::sInstance;
 
-    return true;
-}
+namespace {
 
 static void OnRegister(DNSServiceRef sdRef, DNSServiceFlags flags, DNSServiceErrorType err, const char * name, const char * type,
                        const char * domain, void * context)
 {
-    RegisterContext * sdCtx = reinterpret_cast<RegisterContext *>(context);
-    VerifyOrReturn(CheckForSuccess(sdCtx, __func__, err));
-
     ChipLogDetail(DeviceLayer, "Mdns: %s name: %s, type: %s, domain: %s, flags: %d", __func__, name, type, domain, flags);
 
-    sdCtx->callback(sdCtx->context, type, CHIP_NO_ERROR);
+    auto sdCtx = reinterpret_cast<RegisterContext *>(context);
+    sdCtx->Finalize(err);
 };
 
 CHIP_ERROR Register(void * context, DnssdPublishCallback callback, uint32_t interfaceId, const char * type, const char * name,
-                    uint16_t port, TXTRecordRef * recordRef)
+                    uint16_t port, ScopedTXTRecord & record)
 {
-    DNSServiceErrorType err;
-    DNSServiceRef sdRef;
-    GenericContext * sdCtx = nullptr;
-
-    uint16_t recordLen          = TXTRecordGetLength(recordRef);
-    const void * recordBytesPtr = TXTRecordGetBytesPtr(recordRef);
+    ChipLogProgress(DeviceLayer, "Publishing service %s on port %u with type: %s on interface id: %" PRIu32, name, port, type,
+                    interfaceId);
 
+    GenericContext * sdCtx = nullptr;
     if (CHIP_NO_ERROR == MdnsContexts::GetInstance().GetRegisterType(type, &sdCtx))
     {
-        err = DNSServiceUpdateRecord(sdCtx->serviceRef, nullptr, 0 /* flags */, recordLen, recordBytesPtr, 0 /* ttl */);
-        TXTRecordDeallocate(recordRef);
-        VerifyOrReturnError(CheckForSuccess(sdCtx, __func__, err), CHIP_ERROR_INTERNAL);
+        auto err = DNSServiceUpdateRecord(sdCtx->serviceRef, nullptr, kRegisterFlags, record.size(), record.data(), 0 /* ttl */);
+        VerifyOrReturnError(kDNSServiceErr_NoError == err, sdCtx->Finalize(err));
         return CHIP_NO_ERROR;
     }
 
     sdCtx = chip::Platform::New<RegisterContext>(type, callback, context);
-    err   = DNSServiceRegister(&sdRef, 0 /* flags */, interfaceId, name, type, kLocalDot, nullptr, ntohs(port), recordLen,
-                             recordBytesPtr, OnRegister, sdCtx);
-    TXTRecordDeallocate(recordRef);
-
-    VerifyOrReturnError(CheckForSuccess(sdCtx, __func__, err), CHIP_ERROR_INTERNAL);
+    VerifyOrReturnError(nullptr != sdCtx, CHIP_ERROR_NO_MEMORY);
 
-    err = DNSServiceSetDispatchQueue(sdRef, chip::DeviceLayer::PlatformMgrImpl().GetWorkQueue());
-    VerifyOrReturnError(CheckForSuccess(sdCtx, __func__, err, true), CHIP_ERROR_INTERNAL);
+    DNSServiceRef sdRef;
+    auto err = DNSServiceRegister(&sdRef, kRegisterFlags, interfaceId, name, type, kLocalDot, nullptr, ntohs(port), record.size(),
+                                  record.data(), OnRegister, sdCtx);
+    VerifyOrReturnError(kDNSServiceErr_NoError == err, sdCtx->Finalize(err));
 
     return MdnsContexts::GetInstance().Add(sdCtx, sdRef);
```

Comment:
So if `Add` fails we will _not_ call the callback, looks like?  But on failures above from DNSServiceRegister we will both call the callback _and_ return an error?  Is the caller expecting that behavior?  I would have expected one of error return or call to callback with error, not both....  I'm worried about callers ending up with double-deletes and the like if we do both.

Reply:
None of those cases will call the callback, because "sdCtx" is not "yet" into the list of contexts known by `MdnsContexts::GetInstance`.

Only once "MdnsContexts::GetInstance().Add" has been executed successfully, "Finalize" will start to call the callback.

The previous "CheckForSuccess" method uses to allow the caller to specify if the callback needs to be called but if you look at the current code in master - it is not used in a consistent way - this is why I change the code to makes it implicit / automatically managed".

Reply:
Ah, I had missed that subtlety around Finalize.  OK, thanks!